### ISSUE TITLE
exception:EXC_BAD_INSTRUCTION

### ISSUE THREAD

**Issue by mano233** [OpenWaterAnalytics/EPANET#564]
My environment  is: 
macos(10.15), clion 2019, clang++

I don't know much about c++. I wrote a simple program, but it crashed：
```c++
#include <stdio.h>
#include "epanet2.h"
int  main(int argc, char *argv[]){
  ENopen("example.inp","rtp.txt","");
  return 0;
}
```
Error message:`Exception: EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)`

The final problem is located at input3.c:1175
```c++
strcpy(parser->Tok[0], parser->Tok[1]);
```
It happened `detected source and destination buffer overlap`

how to fix??

**Comment by samhatchett** [comment]
it may help to post the example network you're dealing with - so someone can replicate this behavior.

**Comment by mano233** [comment]
> it may help to post the example network you're dealing with - so someone can replicate this behavior.

ok ,is the [example network](https://drive.google.com/open?id=11srT24Y52JG-QPxI31qZU2tDcb-edDFd)

**Comment by samhatchett** [comment]
Thanks - there does seem to be an issue here.

https://github.com/OpenWaterAnalytics/EPANET/blob/c16c861789f2aa93b7446d08b7a365916c479d2f/src/input3.c#L1174-L1175

... this line offends the latest `strcpy` implementation in Clang because we are trying to copy a character string into a location that might overlap. The `Tok` array is just a set of pointers to the original parsed line, and so there would be a problem depending on the token string lengths.

I can confirm that the implementation in GCC 8.3 does not check for this potential overlap, and so that would not show up on the tests, and might be a workaround if needed.

But - I don't think that this copy operation is necessary. `Tok[0]` is never referenced again for this tokenized line. This same overlap could happen again here, where there's a comment about saving for error reporting, but I can't see where that would ever be reported:

https://github.com/OpenWaterAnalytics/EPANET/blob/c16c861789f2aa93b7446d08b7a365916c479d2f/src/input3.c#L1825-L1828

 Unless I'm missing something, I think these lines are safe to remove.

**Comment by eladsal** [comment]
Is this a C++ thing? I does not fail with the GUI or VB

**Comment by samhatchett** [comment]
no, I think it's just a "c library implementation" thing. GCC works fine, but Clang checks for overlaps.
http://demin.ws/blog/english/2011/07/14/strcpy-on-overlapped-strings/

**Comment by mano233** [comment]
Thanks, I have not noticed the compiler problem.  I changed ”strcpy” to ”strncpy” to solve this problem temporarily, but I don't know what the consequences will be (

**Comment by LRossman** [comment]
The offending line is a carryover from the v2.0 (circa 2000) code. I agree with @samhatchett that with the new method used for reporting errors in an input file it is no longer needed. If there are no objections I will submit a PR that removes the line.

**Comment by LRossman** [comment]
This issue has been resolved via PR #565 .

**Comment by tallpeak** [comment]
strncpy also doesn't allow overlap, in my experience. I hacked my code with this solution, but this should not be used in production code:

#undef strcpy
char *mystrcpy(void *dest, void *source)
{
    size_t sourceLen = strlen(source);
    memcpy(dest, source, sourceLen);
    *((char*)dest + sourceLen) = '\0'; //null terminate
    return dest;
}
#define strcpy mystrcpy

But then I deleted that and used this instead:

void *deleteChar(char *p)
{
    while (*p) {
        *p++ = p[1];
    }
    *p = '\0';
}

Ref: https://github.com/tallpeak/anagram/blob/master/anagramsrt.c
